src/set.c

```diff
@@ -424,3 +424,185 @@ r_obj* vec_set_union(r_obj* x,
+
+  return vec_set_symmetric_difference(x, y, ptype, &x_arg, &y_arg, call);
+}
+
+r_obj* vec_set_symmetric_difference(r_obj* x,
+                                    r_obj* y,
+                                    r_obj* ptype,
+                                    struct vctrs_arg* x_arg,
+                                    struct vctrs_arg* y_arg,
+                                    struct r_lazy call) {
+  int n_prot = 0;
+
+  if (ptype == r_null) {
+    int _;
+
+    ptype = vec_ptype2_params(
+      x,
+      y,
+      x_arg,
+      y_arg,
+      call,
+      &_
+    );
+    KEEP_N(ptype, &n_prot);
+
+    ptype = vec_ptype_finalise(ptype);
+    KEEP_N(ptype, &n_prot);
+  }
+
+  x = vec_cast_params(
+    x,
+    ptype,
+    x_arg,
+    vec_args.empty,
+    call,
+    S3_FALLBACK_false
+  );
+  KEEP_N(x, &n_prot);
+
+  y = vec_cast_params(
+    y,
+    ptype,
+    y_arg,
+    vec_args.empty,
+    call,
+    S3_FALLBACK_false
+  );
+  KEEP_N(y, &n_prot);
+
+  r_obj* x_proxy = KEEP_N(vec_proxy_equal(x), &n_prot);
+  x_proxy = KEEP_N(vec_normalize_encoding(x_proxy), &n_prot);
+
+  r_obj* y_proxy = KEEP_N(vec_proxy_equal(y), &n_prot);
+  y_proxy = KEEP_N(vec_normalize_encoding(y_proxy), &n_prot);
+
+  const r_ssize x_size = vec_size(x_proxy);
+  const r_ssize y_size = vec_size(y_proxy);
+
+  struct dictionary* x_dict = new_dictionary(x_proxy);
+  PROTECT_DICT(x_dict, &n_prot);
+
+  struct dictionary* y_dict = new_dictionary(y_proxy);
+  PROTECT_DICT(y_dict, &n_prot);
+
+  r_obj* x_marked_shelter = KEEP_N(r_alloc_raw(x_size * sizeof(bool)), &n_prot);
+  bool* v_x_marked = (bool*) r_raw_begin(x_marked_shelter);
+
+  r_obj* y_marked_shelter = KEEP_N(r_alloc_raw(y_size * sizeof(bool)), &n_prot);
+  bool* v_y_marked = (bool*) r_raw_begin(y_marked_shelter);
+
+  // Load dictionary with `x`.
+  // Key values point to first time we saw that `x` value.
+  // Mark those first seen locations as possible results.
+  for (r_ssize i = 0; i < x_size; ++i) {
+    const uint32_t hash = dict_hash_scalar(x_dict, i);
+    const bool first_time = x_dict->key[hash] == DICT_EMPTY;
+
+    if (first_time) {
+      dict_put(x_dict, hash, i);
+    }
+
+    v_x_marked[i] = first_time;
+  }
+
+  // Load dictionary with `y`.
+  // Key values point to first time we saw that `y` value.
+  // Mark those first seen locations as possible results.
+  for (r_ssize i = 0; i < y_size; ++i) {
+    const uint32_t hash = dict_hash_scalar(y_dict, i);
+    const bool first_time = y_dict->key[hash] == DICT_EMPTY;
+
+    if (first_time) {
+      dict_put(y_dict, hash, i);
+    }
+
+    v_y_marked[i] = first_time;
+  }
+
+  r_ssize n_x_marked = x_dict->used;
+  r_ssize n_y_marked = y_dict->used;
+
+  // Check if unique elements of `y` are in `x`.
+  // If they are, unmark them from both `x` and `y`.
+  for (r_ssize i = 0; i < y_size; ++i) {
+    if (!v_y_marked[i]) {
+      continue;
+    }
+
+    const uint32_t hash = dict_hash_with(x_dict, y_dict, i);
```

Comment:
I'm always confused by this operation. I wish we could find a better name for it.